src/btree/bt_sync.c

```diff
@@ -9,6 +9,66 @@
 #include "wt_internal.h"
 
 /*
+ * __sync_checkpoint_can_skip --
+ *	There are limited conditions under which we can skip writing a dirty
+ * page during checkpoint.
+ */
+static inline bool
+__sync_checkpoint_can_skip(WT_SESSION_IMPL *session, WT_PAGE *page)
+{
+	WT_PAGE_MODIFY *mod;
+	WT_MULTI *multi;
+	WT_TXN *txn;
+	u_int i;
+
+	mod = page->modify;
+	txn = &session->txn;
+
+	/*
+	 * We can skip some dirty pages during a checkpoint. The requirements:
+	 *
+	 * 1. they must be leaf pages,
+	 * 2. there is a snapshot transaction active (which is the case in
+	 *    ordinary application checkpoints but not all internal cases),
+	 * 3. the first dirty update on the page is sufficiently recent the
+	 *    checkpoint transaction would skip them,
+	 * 4. there's already an address for every disk block involved.
+	 */
+	if (WT_PAGE_IS_INTERNAL(page))
+		return (false);
+	if (!F_ISSET(txn, WT_TXN_HAS_SNAPSHOT))
+		return (false);
+	if (!WT_TXNID_LT(txn->snap_max, mod->first_dirty_txn))
+		return (false);
+
+	/*
+	 * The problematic case is when a page was evicted but when there were
+	 * unresolved updates and not every block associated with the page has
+	 * a disk address. We can't skip such pages because we need a checkpoint
+	 * write with valid addresses.
+	 *
+	 * The page's modification information can change underfoot if the page
+	 * is being reconciled, so we'd normally serialize with reconciliation
+	 * before reviewing page-modification information. However, checkpoint
+	 * is the only valid writer of dirty leaf pages at this point, we skip
+	 * the lock.
+	 */
+	if (mod->rec_result == WT_PM_REC_MULTIBLOCK)
+		for (multi = mod->mod_multi,
+		    i = 0; i < mod->mod_multi_entries; ++multi, ++i)
+			if (multi->addr.addr == NULL)
+				return (false);
+
+	/*
+	 * Mark the tree dirty if skipping a page, the checkpoint marked it
+	 * clean and we can't skip future checkpoints until this page is
+	 * written.
+	 */
+	__wt_tree_modify_set(session);
```

Comment:
Would you mind pull this up into the caller so the inlined function has no side effects?